pandas/core/indexes/base.py

```diff
@@ -5178,7 +5178,16 @@ def _maybe_promote(self, other: Index):
         if we can upcast the object-dtype one to improve performance.
         """
 
-        if self.inferred_type == "date" and isinstance(other, ABCDatetimeIndex):
+        if isinstance(self, ABCDatetimeIndex) and isinstance(other, ABCDatetimeIndex):
```

Comment:
why should this not raise?

Reply:
This is relevant in three places: get_indexer, get_indexer_non_unique, and join.  In the first two, we're talking about comparisons, which always allow mixed tz.  In the third, that is just The Way It Is, xref #39328

Reply:
ok i see, yeah generally i find this weird that we allow indexing on mixed tz's but i guess that boat has sailed.